### PR TITLE
changes to setup, remove teal

### PR DESCRIPTION
**Pull request by cshanahan1**
Made changes to setup files to install dependencies and remove 'relic' I also removed some pytest setup since there are currently no tests in this package. I also removed everything related to 'teal'.

### CODE REVIEW

**Review by pllim — CHANGES_REQUESTED** [review]
I am not sure if this is enough. I don't see any `pyproject.toml` . Also there are still metadata left in `setup.py` that can be moved to `setup.cfg`.

You also want to try to actually package this thing and make sure you can install both the packaged tarball and wheel (separately) and get back a working install. If you want to take this a step further, you can automate like https://github.com/astropy/pytest-doctestplus/blob/main/.github/workflows/publish.yml .

It is unfortunate that this package does not have a test suite, so all you can do is to make sure the installation works (i.e., you can import the package) but not guarantee that it would work correctly.

**Comment by pllim** [comment]
p.s. I see Matt R put in https://github.com/spacetelescope/wfc3tools/blob/master/.github/workflows/publish-to-pypi.yml but that does not run as CI and not very sure how you ask it to only run as packaging test but not upload anything. Anyways, you cannot have two publish workflows, so you need to delete this one if you are going to add the other one to CI.

**Comment by nden** [comment]
As Pey-Lian mentioned - add a pyproject.toml file. Then refactor `__init__.py` to define `__version__` similar to other packages.

**Review by FDauphin — CHANGES_REQUESTED** [review]
In `setup.py`, we should determine what information goes as PACKAGENAME, AUTHOR, etc.

In `__init__.py`, if we're going to use `from __future__` that should be refactored. `from .version` is looking for a module named `wfc3tools.version`, which doesn't exist. Nadia mentioned refactoring `__version__` like the other packages, but I don't know if that solves the problem.

**Comment by pllim** [comment]
@FDauphin , all the metadata should now be in `setup.cfg`. And `__version__` is generated by `setuptools_scm`. You don't need `from __future__`.

File: MANIFEST.in [hunk 1]
```diff
@@ -1,6 +1,7 @@
-include RELIC-INFO
 include README.rst
 include CHANGES.rst
+include setup.cfg
+include LICENSE
```
Comment (pllim) [inline MANIFEST.in:4]:
```suggestion
include LICENSE
include pyproject.toml
```

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+[build-system]
+requires = ["setuptools>=42", "setuptools_scm[toml]>=3.4", "wheel"]
+build-backend = "setuptools.build_meta"
+
+[tool.setuptools_scm]
```
Comment (pllim) [inline pyproject.toml:5]: Don't need this?
```suggestion
```

File: setup.cfg [hunk 3]
```diff
@@ -15,6 +15,25 @@ project_urls =
...
+    matplotlib>=3.5.1
+    stsci.tools>=4.0.1
+    scipy>=1.8.0
+
+[options.extras_require]
+docs =
+    sphinx
+    sphinx-automodapi
+    sphinx-rtd-theme
+    stsci-rtd-theme
```
Comment (pllim) [inline setup.cfg:34]: Do you really need two different themes? Pick one that you actually are using in `docs/conf.py`.

File: setup.cfg [hunk 4]
```diff
@@ -15,6 +15,25 @@ project_urls =
...
+    sphinx-astropy
+
 [build-sphinx]
```
Comment (pllim) [inline setup.cfg:37]: I don't think you need `build-sphinx` either. Do you use it?

File: wfc3tools/__init__.py [hunk 5]
```diff
@@ -32,3 +34,12 @@
     teal.print_tasknames(__name__, os.path.dirname(__file__))
 except (ImportError, UnboundLocalError):
     print("Teal not installed, gui param editing disabled")
+
+if sys.version_info < (3, 6):
+    raise ImportError("wfc3tools supports Python versions 3.6 and above.")  # pragma: no cover
+
```
Comment (pllim) [inline wfc3tools/__init__.py:40]: Don't need this. Enforced by setting in `setup.cfg`.
```suggestion
```

File: wfc3tools/__init__.py [hunk 6]
```diff
@@ -32,3 +34,12 @@
...
+try:
+    __version__ = get_distribution(__name__).version
```
Comment (pllim) [inline wfc3tools/__init__.py:42]: This is overkill. If setuptools_scm worked correctly, you can do `from .version import version as __version__` directly.

**Review by pllim — COMMENTED** [review]
(Just a quick review of that you have so far.)

Change `package_name` in `setup.cfg` to just `name`.

You don't need anything in `setup.py` besides invoking `setuptools_scm` to generate the version file. Move all the remaining metadata to appropriate sections in `setup.cfg`.

File: setup.cfg (hunk 4 above)
Comment (cshanahan1) [reply setup.cfg:37]: i'm not sure, i'll take it out for now and i'm going to be making another PR with some changes for documentation

File: setup.cfg (hunk 3 above)
Comment (cshanahan1) [reply setup.cfg:34]: i included both because jwst and romancal have both themes included, I wasn't sure which one was needed to make the docs look like both of those packages

File: wfc3tools/__init__.py (hunk 6 above)
Comment (cshanahan1) [reply wfc3tools/__init__.py:42]: @nden commented to say I should add __version__ in __init__

File: pyproject.toml (hunk 2 above)
Comment (cshanahan1) [reply pyproject.toml:5]: roman and jwst both include this in this file, is it not necessary? what does including it here imply, that its needed for the setup?

File: setup.cfg (hunk 3 above)
Comment (cshanahan1) [reply setup.cfg]: removed both for now, will add back in docs options in another PR

File: wfc3tools/__init__.py (hunk 6 above)
Comment (cshanahan1) [reply wfc3tools/__init__.py]: i've remove the minimum version check in this file, since setup.cfg should be enforcing this (i think?). i also did not add a __version__ here because even without declaring it here, I am still able to access wfc3tools.__version__, so I assume this is done by setuptools

File: setup.py [hunk 7]
```diff
@@ -1,82 +1,17 @@
...
-        TestCommand.finalize_options(self)
-        self.test_args = ['wfc3tools/tests']
-        self.test_suite = True
-
-    def run_tests(self):
-        # import here, cause outside the eggs aren't loaded
-        import pytest
-        errno = pytest.main(self.test_args)
-        sys.exit(errno)
+NAME = 'wfc3tools'
```
Comment (nden) [inline setup.py:4]: I think this is not needed. It's defined in setup.cfg

File: setup.cfg [hunk 8]
```diff
@@ -14,24 +13,21 @@ project_urls =
...
+[options]
+python_requires = >=3.8
+setup_requires =
+    setuptools_scm
+install_requires =
+    astropy>=4.2.0
+    numpy>=1.21.0
+    matplotlib>=3.5.1
+    stsci.tools>=4.0.1
+    scipy>=1.8.0
```
Comment (nden) [inline setup.cfg:33]: I would add also 
`zip_safe=False` 
in this section.

File: MANIFEST.in [hunk 9]
```diff
@@ -1,6 +1,8 @@
...
```
Comment (nden) [inline MANIFEST.in:4]: The actual name is `LICENSE.txt`

**Review by nden — CHANGES_REQUESTED** [review]
I installed the package and found a few minor issues. After fixing them I think this is good to go.

File: setup.cfg [hunk 10]
```diff
@@ -1,37 +1,34 @@
...
-norecursedirs = build docs/_build relic
-
-[entry_points]
-
-[bdist_wheel]
-# This flag says that the code is written to work on both Python 2 and Python
-# 3. If at all possible, it is good practice to do this. If you cannot, you
-# will need to generate wheels for each Python version that you support.
-universal=1
+[options]
```
Comment (pllim) [inline setup.cfg:24]:
```suggestion
[options]
packages = find:
```

File: setup.py [hunk 11]
```diff
@@ -1,82 +1,7 @@
...
-        'astropy',
-    ],
-    tests_require=['pytest'],
-    packages=find_packages(),
-    package_data={PACKAGENAME: ['pars/*']},
-    cmdclass = {'test': PyTest}
-)
+    use_scm_version=True,
+    setup_requires=['setuptools_scm'],
+    packages=find_packages())
```
Comment (pllim) [inline setup.py:7]:
```suggestion
    use_scm_version={'write_to': 'wfc3tools/version.py'})
```

File: setup.py [hunk 12]
```diff
@@ -1,82 +1,7 @@
 #!/usr/bin/env python
-import os
-import subprocess
-import sys
 from setuptools import setup, find_packages
```
Comment (pllim) [inline setup.py:2]:
```suggestion
from setuptools import setup
```

File: wfc3tools/__init__.py [hunk 13]
```diff
@@ -21,14 +24,5 @@
...
-import os
 
-# These lines allow TEAL to print out the names of TEAL-enabled tasks
-# upon importing this package.
-try:
-    from stsci.tools import teal
-    teal.print_tasknames(__name__, os.path.dirname(__file__))
-except (ImportError, UnboundLocalError):
-    print("Teal not installed, gui param editing disabled")
+__version__ = version('wfc3tools')
```
Comment (pllim) [inline wfc3tools/__init__.py:28]:
```suggestion
try:
    from .version import version as __version__
except ImportError:
    __version__ = ''
```

File: wfc3tools/__init__.py [hunk 14]
```diff
@@ -7,7 +7,10 @@
 
 
 """
-from __future__ import absolute_import, print_function
+import os
+import sys
+
+from importlib.metadata import version
```
Comment (pllim) [inline wfc3tools/__init__.py:13]:
```suggestion
```

File: wfc3tools/calwf3.py [hunk 15]
```diff
@@ -1,21 +1,11 @@
...
 import os.path
 import subprocess
-from .version import __version_date__, __version__
 
 # STSCI
 from stsci.tools import parseinput
 from .util import error_code
 
-try:
-    from stsci.tools import teal
```
Comment (pllim) [inline wfc3tools/calwf3.py:13]: Is WFC3 team ok with removing TEAL support? This seems out of scope.

**Review by pllim — CHANGES_REQUESTED** [review]
Close...

If you want to remove the TEAL doc, should at least move them to proper RTD sections so we don't lose the info.

Also, I don't see any CI set up to test the packaging.

File: wfc3tools/calwf3.py (hunk 15 above)
Comment (cshanahan1) [reply wfc3tools/calwf3.py:13]: @FDauphin

Comment (FDauphin) [reply wfc3tools/calwf3.py:13]: Back in September, I asked Michele De La Pena (might be helpful to ask her review as well) if we could remove TEAL support, but she said she was unsure. I am in support of removing it because I do not know in any instances where TEAL is the preferred method of using `wfc3tools`.

**Review by FDauphin — APPROVED** [review]
I was able to clone this branch, install `wfc3tools`, and import with no issues. With regards to TEAL, I asked Michele De La Pena if removing it is the right decision (which I think it is).

File: wfc3tools/calwf3.py (hunk 15 above)
Comment (pllim) [reply wfc3tools/calwf3.py:13]: I agree that TEAL has probably outlived its usefulness, but someone should do the due diligence to make sure all the doc you are deleting are already reflected in RTD docs. If not, they have to be ported. Otherwise, you might lose precious info. Up to you.

Comment (FDauphin) [reply wfc3tools/calwf3.py:13]: @cshanahan1 is correct; the help files info is added as docstrings in PEP 8 style and is added in the RTD. However, the new RTD I initially removed all TEAL dependency. I suppose it would be best to add a sentence or two on the main RTD page that TEAL dependency was removed and why so.

**Comment by nden** [comment]
I built and tested a source package and a wheel. I can confirm they work.

File: setup.py (hunk 11 above)
Comment (cshanahan1) [reply setup.py]: done, but just curious why 'packages' isn't needed?

File: setup.cfg (hunk 10 above)
Comment (cshanahan1) [reply setup.cfg:24]: oh does this replace find_packages() in setup.py? im confused about what belongs in this file vs that file, i had it in .py because thats what i've seen elsewhere.

**Comment by nden** [comment]
@pllim The idea for this PR is to make the build scripts up to date and release so we can start putting together and testing a larger HST distribution. The documentation will come in the next PR, followed by a new release but meanwhile we can work on the larger HST distribution. Is there anything else you would like to see done for this PR?

File: setup.cfg (hunk 10 above)
Comment (pllim) [reply setup.cfg:24]: Yes, it is a different way to say the same thing.

**Review by pllim — APPROVED** [review]
For completeness, maybe remove the following from `.gitignore`:

```
RELIC-INFO
relic/
```

Otherwise, LGTM. Thanks!

**Comment by cshanahan1** [comment]
ok thanks everyone! im going to merge this now

**Comment by FDauphin** [comment]
@cshanahan1 thanks for your work on this! Michele followed up and said as long as `drizzlepac` (which still uses TEAL) isn't used in `wfc3tools`, we can remove TEAL. `wfc3tools` doesn't use `drizzlepac` so removing TEAL was the right move.
